keps/104-StartupPolicy/README.md

```diff
@@ -0,0 +1,420 @@
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+  - [Poor Person Workflow Engine](#poor-person-workflow-engine)
+  - [Increase number of patches to jobset](#increase-number-of-patches-to-jobset)
+- [Alternatives](#alternatives)
+  - [Operate on Conditions](#operate-on-conditions)
+  - [Allow for both ready and succeeded](#allow-for-both-ready-and-succeeded)
+<!-- /toc -->
+
+## Summary
+
+This KEP adds a StartupPolicy for Jobsets.  A StartupPolicy allows for some flexibility on when to start ReplicatedJobs in a Jobset.  
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+## Motivation
+
+High Performance Computing / Machine Learning users usually want some control over how jobs are being created.  A Startup policy would allow a user to specify which status a replicated job should have before starting other replicated jobs in a jobset.
+
+### Goals
+
+- Add a startup policy for JobSet
+
+### Non-Goals
+
+- Complicated workflow specifications
+  - This means that we will support simple dependecies rather than processing a full dag.
+  - Projects like argo-workflows are a better use case for this.
+- Starting Jobs based on Ready will be at the ReplicatedJob level not at the Job level.
+  - ReplicatedJobs will be classified as ready if all jobs are ready.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, I have one ReplicatedJob that functions as the driver and a series of worker ReplicatedJobs.  
+I want the driver to be ready before I start the worker replicated jobs.  
+This is useful for HPC/AI/ML workloads as some frameworks have a driver pod that must be ready before the workers are started.  
+
+The API below would fit my purpose.
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: driver-ready-worker-start
+spec:
+  startupPolicy:
+     replicatedJobsStartInOrder: true
+  replicatedJobs:
+  - name: driver
+    replicas: 1
+    template:
```

Comment:
stupid question - this is just a Job specification right used as the template? https://github.com/kubernetes-sigs/jobset/blob/4222b34572b8cd8b8a1a7863c9fde5098bd370a3/api/jobset/v1alpha2/jobset_types.go#L136 

So why can't we put the readiness logic in there instead?

Reply:
I don't really understand what you are asking.  Technically the readiness logic would go there but in the pod template of the JobTemplate (turtles all the way down).  

Are you thinking more that there should be a JobReadiness field?

Reply:
I'm generally thinking how we can simplify the design to put some of the logic into Job, and kill two birds with one stone. It's fairly abstract thinking at the moment and I didn't do a good job of articulating myself (and probably need to think more on this before it's ready for that).